storage/client_range_tree_test.go

```diff
@@ -18,668 +18,180 @@
+	if node.RightKey != nil {
+		loadNodes(t, db, *node.RightKey, nodes)
 	}
-	return nil
 }
 
-// treeNodesEqual compares the expectedTree from the provided key to the actual
-// nodes retrieved from the db.  It recursively calls itself on both left and
-// right children if they exist.
-func treeNodesEqual(db *client.DB, expected testRangeTree, key proto.Key) error {
-	expectedNode, ok := expected.Nodes[string(key)]
-	if !ok {
-		return util.Errorf("Expected does not contain a node for %s", key)
-	}
-	actualNode := &proto.RangeTreeNode{}
-	if err := db.GetProto(keys.RangeTreeNodeKey(key), actualNode); err != nil {
-		return err
+// loadTree loads the tree root and all of its nodes. It puts all of the nodes
+// into a map.
+func loadTree(t *testing.T, db *client.DB) (*proto.RangeTree, map[string]proto.RangeTreeNode) {
+	tree := new(proto.RangeTree)
+	if err := db.GetProto(keys.RangeTreeRoot, tree); err != nil {
+		t.Fatal(err)
 	}
-	if err := nodesEqual(key, expectedNode, *actualNode); err != nil {
-		return err
+	nodes := make(map[string]proto.RangeTreeNode)
+	if tree.RootKey != nil {
+		loadNodes(t, db, tree.RootKey, nodes)
 	}
-	if expectedNode.LeftKey != nil {
-		if err := treeNodesEqual(db, expected, *expectedNode.LeftKey); err != nil {
-			return err
-		}
-	}
-	if expectedNode.RightKey != nil {
-		if err := treeNodesEqual(db, expected, *expectedNode.RightKey); err != nil {
-			return err
-		}
-	}
-	return nil
+	return tree, nodes
 }
 
-// treesEqual compares the expectedTree and expectedNodes to the actual range
-// tree stored in the db.
-func treesEqual(db *client.DB, expected testRangeTree) error {
-	// Compare the tree roots.
-	actualTree := &proto.RangeTree{}
-	if err := db.GetProto(keys.RangeTreeRoot, actualTree); err != nil {
-		return err
-	}
-	if !reflect.DeepEqual(&expected.Tree, actualTree) {
-		return util.Errorf("Range tree root is not as expected - expected:%+v - actual:%+v", expected.Tree, actualTree)
+// VerifyTree checks to ensure that the tree is indeed balanced and a correct
+// red-black tree. It does so by checking each of the red-black tree properties.
+// These verify functions are similar to the those found in the range_tree_test
+// but these use a map of nodes instead of a tree context.
+func VerifyTree(t *testing.T, tree *proto.RangeTree, nodes map[string]proto.RangeTreeNode, testName string) {
+	root, ok := nodes[tree.RootKey.String()]
+	if !ok {
+		t.Fatalf("%s: could not find root node with key %s", testName, tree.RootKey)
 	}
 
-	return treeNodesEqual(db, expected, expected.Tree.RootKey)
+	// Property 1 is always correct. All nodes are already colored.
+	verifyProperty2(t, testName, &root)
+	// Property 3 is always correct. All leaves are black.
+	verifyProperty4(t, nodes, testName, &root)
+	pathBlackCount := new(int)
+	*pathBlackCount = -1
+	verifyProperty5(t, nodes, testName, &root, 0, pathBlackCount)
```

Comment:
There's no actual check here that the tree is sorted, just that node relationships obey red/black rules.